draft-ietf-ppm-dap.md

```diff
@@ -1323,26 +1323,57 @@ the input share is marked as invalid with a corresponding ReportShareError
-   parameter. If this check fails, the input share MUST be marked as invalid
-   with the error `report_replayed`. This is the case if the report was used in
-   a previous aggregate request and is therefore a replay.
+1. Check if the report is too far into the future. Implementors can provide for
+   some small leeway, usually no more than a few minutes, to account for clock
+   skew. If a report is rejected for this reason, the Aggregator SHOULD mark the
+   input share as invalid with the error `report_too_early`.
+
+1. Check if the report's timestamp has passed its task's `task_expiration` time,
+   if so the Aggregator MAY mark the input share as invalid with the error
+   "task_expired".
+
+1. Check if the PlaintextInputShare contains unrecognized extensions. If so,
+   then the Aggregator MUST mark the input share as invalid with error
+   "unrecognized_message".
+
+1. Check if the ExtensionType of any two extensions in PlaintextInputShare is
+   the same. If so, then the Aggregator MUST mark the input share as invalid
+   with error "unrecognized_message".
+
+1. Check if the report may still be aggregated with the current aggregation
+   parameter. This can be done by looking up all aggregation parameters
+   previously used for this report and calling
+
+   ~~~
+   VDAF.is_valid(current_agg_param, previous_agg_params)
+   ~~~
+
+   If this returns false, the input share MUST be marked as invalid with the
+   error `report_replayed`.
+
     * Implementation note: To detect replay attacks, each Aggregator is required
       to keep track of the set of reports it has processed for a given task.
       Because honest Clients choose the report ID at random, it is sufficient to
       store the set of IDs of processed reports. However, implementations may
       find it helpful to track additional information, like the timestamp, so
       that the storage used for anti-replay can be sharded efficiently.
 
-1. Check if the report has never been aggregated but is contained by a batch
-   that has been collected. If this check fails, the input share MUST be marked
-   as invalid with the error `batch_collected`. This prevents additional reports
-   from being aggregated after its batch has already been collected.
+1. Some VDAFs allow a batch to be collected multiple times; the aggregators must
+   check that the exact same set of reports is being aggregated each time.
+   A report may not be aggregated into a batch unless it was included in all
+   previous collections for the batch. Checking that no batches are omitted
+   from a batch is covered in {{batch-validation}}.
+   If this check fails, the input share MUST be marked as invalid with the error
+   `batch_collected`. This prevents collectors from learning anything about
+   small numbers of reports that are uploaded between two collections of a
+   batch.
+   The helper must consider a batch as collected when it has sent the
+   `AggregateShare` to the leader.
```

Comment:
nit: IMO this doesn't need to be clarified here.
```suggestion
```

Reply:
I agree with Chris, but also agree with Simon that DAP should be more clear about the lifecycle or state machine of various objects in the protocol, because otherwise it's kinda hard for implementors to figure out stuff like when a helper aggregate share is considered to be collected. I'd like to have a section in DAP where we discuss all that, so that then other parts of the protocol like this could refer to it, and then it'd be clear just what we mean when we say that a report is aggregated or a batch is collected. We talked about that a bit in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/384. In any case, I certainly don't want to solve that problem in this PR. Maybe we could take Simon's text and include a TODO referencing #384?